[4-9단계 - 방탈출 예약 관리] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
## 코드 리뷰 팁

## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
https://prolog.techcourse.co.kr/studylogs/4153

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->
안녕하세요 샐리~!
이번 리뷰도 잘부탁드립니다.😊

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

kjyyjk committed:
feat: h2 db 연동을 위한 application.properties 속성 추가

- spring.sql.init.mode=always
Spring에서 제공하는 스크립트 기반 sql 초기화를 사용하기 위해 추가해야하는 속성이다.
ex) data.sql, schema.sql
h2를 사용하는 경우 이 속성이 기본으로 설정되어 있기 때문에 따로 추가해줄 필요는 없으나,
이후 db를 변경하는 경우 발생할 예외를 방지하기 위해 명시해주었다.

- spring.h2.console.enabled=true
h2에는 gui 콘솔이 내장되어 있는데 spring에서는 기본적으로 비활성화 되어있다.
따라서 콘솔 사용을 위해 활성화한다.

- spring.datasource.url=jdbc:h2:mem:database
mem은 인메모리 db를 사용한다는 의미이며 database는 사용할 db의 이름이다.
인메모리 db는 대부분의 데이터를 디스크가 아닌 메모리에 저장하는 db이다.
디스크 io작업을 최소화하거나 제거하고 메모리에서 바로 데이터에 액세스하기 때문에
속도가 빠르다는 단점이 있다. 다만 db 서버의 전원이 꺼진다면 메모리에 있는 데이터가
모두 소실될 수 있다. 그렇기 때문에 애플리케이션 개발 중 테스트 db로 자주 사용한다.

kjyyjk committed:
feat: reservation 테이블을 생성하는 schema.sql 작성

kjyyjk committed:
refactor: Reservations 생성

- 기존 코드는 Controller에 예약들을 관리하는 책임이 존재했다.
하지만 내가 생각한 Controller는 요청에 따른 도메인 로직 호출 및
요청 처리 결과에 따른 응답 구성을 하며, 클라이언트와 서버 사이 중개 책임을 가진다.
즉 현재 Controller는 두 개 이상의 큰 책임들을 가지고 있고,
어느 한 부분에 변경 발생 시 다른 로직까지 변경이 전파되기가 쉽다는 문제를 가진다.
따라서 예약들을 저장하고 관리하는 Reservation을 생성하여 책임을 분리한다.

- 기존에는 RequestDto를 Reservation을 변환할 때 이미 Controller가
다음 id 값을 알고 있었다. 하지만 리팩터링 이후에는 Controller가 dto변환 시
id값을 넣어줄 수가 없다. 따라서 Reservation id값을 래퍼 클래스로 변경해,
id에 null 값이 들어올 수 있도록 허용해주었다. 이후 저장 시 Reservations 내부에서
새로운 Reservation을 만들면서 id값을 설정해준다.

kjyyjk committed:
feat: Jdbc를 통해 DB와 통신하는 ReservationDao 생성

- JdbcTemplate은 스프링에서 자동으로 구성해주며 thread-safe하다는
특징이 있다. 따라서 구현한 코드와 같이 스프링 빈으로 주입받아 사용할 수 있다.

- 스프링 빈 주입을 받기 위해 ReservationDto 클래스도 @Component 애너테이션을
달아서 스프링 빈으로 만든다.

kjyyjk committed:
refactor: ReservationDao 인터페이스화

- 미션 요구사항에 따라 예약 조회 시에만 jdbc를 활용해야하므로
우선 Controller가 기존 List 기반/jdbc 기반 Dao를 모두 가지도록 한다.

- jdbc 기반 Dao의 주입이 필요하기 때문에 단위 테스트가 깨지는 현상을 발견할 수 있다.
jdbc 통신을 하는 것은 단위 테스트가 아니라고 판단했기 때문에 단위 테스트에서는
기존 List 기반 Dao를 넣어주고자 했고 이를 위해 두 Dao를 추상화 해서 인터페이스를 생성했다.

- ReservationDao라는 역할의 책임은 예약을 보관하고 관리하는 것이다.
ListBasedReservationDao 객체의 책임은 List 기반으로 예약을 보관하고 관리하는 것이다.
JdbcBasedReservationDao 객체의 책임은 jdbc를 사용해 예약을 보관하고 관리하는 것이다.

kjyyjk committed:
refactor: 패키지 구조 변경

kjyyjk committed:
feat: db를 활용해 예약 추가 구현

kjyyjk committed:
feat: db를 활용해 예약 취소 구현

kjyyjk committed:
fix: KeyHolder가 제대로 생성 키를 받아오도록 수정

kjyyjk committed:
test: 6단계 api 테스트 추가

kjyyjk committed:
fix: step_five 테스트가 통과하도록 수정

- 응답 본문을 Reservation이 아닌 ReservationResponse로 역직렬화하게 수정한다.

kjyyjk committed:
refactor: 잔여 코드 제거

kjyyjk committed:
docs: README에 API 명세

kjyyjk committed:
docs: 방탈출 시간관리 기능목록 추가

kjyyjk committed:
feat: 방탈출 시간 관리 페이지 응답 api 구현

kjyyjk committed:
feat: reservation_time 스키마 추가

kjyyjk committed:
feat: 방탈출 시간 조회 api 구현

kjyyjk committed:
feat: 방탈출 시간 추가 api 구현

kjyyjk committed:
feat: 방탈출 시간 삭제 api 구현

kjyyjk committed:
test: 7단계 api 테스트 추가

kjyyjk committed:
test: ReservationTimeController 단위 테스트 수행

- 단위 테스트를 위해 ReservationTimeDao 인터페이스로 추상화하고
테스트 시 FakeReservationTimeDao 구현체를 주입해주었다.

kjyyjk committed:
refactor: ReservationTimeController에서 ResponseEntity를 반환하도록 통일

- 통일하는 이유는 다음과 같다.
일관성을 지킬 수 있다. 필요 시 헤더와 상태 코드를 간편하게 지정할 수 있다.
응답 코드가 명확하다.

kjyyjk committed:
refactor: Dao의 `@Component`를 `@Repository`로 변경

kjyyjk committed:
refactor: 예약과 예약 시간 테이블의 스키마 수정

- 예약 테이블이 예약 시간 테이블의 id를 외래키로 가진다.

kjyyjk committed:
feat: ReservationTimeDao#getById 추가 구현

- 미션 요구사항에 따라 Reservation과 ReservationTime이 연관관계를 가져야한다.
Reservation을 저장할 때 ReservationTime이 필요하므로,
id 기반 ReservationTime 조회를 구현한다.

kjyyjk committed:
refactor: GET /admin/reservation 응답 페이지 변경

kjyyjk committed:
refactor: Reservation이 ReservationTime에 의존하게 리팩터링

- 예약 조회 시 inner join을 통해 연관된 reservationTime 테이블까지 함께 조회한 뒤,
Reservation을 조립해 반환한다.

- 예약 추가 시 사용자로부터 받은 ReservationTime의 id를 기반으로 ReservationTime을 조회한 뒤
Reservation을 생성해서 추가한다. 이를 위해 ReservationController에서 ReservationTimeDao를
생성자 주입 받는다.

kjyyjk committed:
refactor: ReservationTimeController에 url prefix 설정

kjyyjk committed:
test: 8단계 api 테스트 추가

kjyyjk committed:
refactor: time.dao 패키지 분리

kjyyjk committed:
test: 9단계 api 테스트 추가

kjyyjk committed:
style: 불필요한 개행 제거

kjyyjk committed:
docs: 예약 시간 추가로 인한 API 명세 최신화

kjyyjk commented on src/main/java/roomescape/reservation/web/ReservationController.java:
```diff
@@ -11,32 +9,42 @@
-        Reservation reservation = request.toReservation(index.getAndIncrement());
-        reservations.add(reservation);
-        ReservationResponse response = ReservationResponse.from(reservation);
+        ReservationTime time = reservationTimeDao.getById(request.timeId());
+        Reservation savedReservation = reservationDao.save(request.toReservation(time));
+        ReservationResponse response = ReservationResponse.from(savedReservation);
```
이번 미션에서는 레이어드 아키텍처를 적용해볼 것을 요구했습니다.

이전 단계부터 DB에 접근하는 것은 Controller의 책임이 아니라고 판단했기 때문에 Repository 레이어는 큰 고민 없이 분리했습니다. 하지만 Service 레이어에 대해서는 많이 고민 됩니다..!

제가 생각하는 Service 레이어의 책임은 `비즈니스 로직 수행`인데요, `비즈니스 로직이 무엇인가`에 대한 저의 기준이 아직 모호한 것 같아요. 그나마 현재 코드에서 가장 비즈니스 로직스럽다고 생각되는 부분이 이 부분입니다.
그러나 이처럼 여러 Dao에 의존해서 하나의 도메인 객체를 생성하는 것이 과연 비즈니스 로직인가에 대한 확신이 없습니다.

궁금증을 해소하기 위해 Service 레이어를 분리한 크루들에게 근거를 물어보았고,
대부분의 답변은 다음과 같았습니다.
- 컨트롤러가 도메인 객체이자 엔티티인 Reservation과 ReservationTime을 알고 있는 것은 책임을 벗어난 것이다.
- 엔티티는 민감한 정보인데 뷰와 가장 근접한 컨트롤러까지 내려주는 것이 불편하다.

이 의견들에 저도 일부 동의하고 추후에 예외 처리가 추가되거나 하면 결국 Service 레이어를 분리할 것 같습니다.
하지만 현재 상황에서는 아직 그 필요성이 크게 느껴지지 않는 것 같아요.

그래서 우선 Service 레이어를 분리하지는 않았습니다.

dusdn1702 replied:
지금의 미션에서 요구사항이 많지 않아 Service 레이어에 대한 필요성이 없다라는 의견에 동의해요.
다만 코멘트 남긴 것처럼, Service 대신 Dao와 Dto를 사용하신 만큼 각 객체들이 어떠한 책임을 가지고 있는지에 대해 이야기 나누어보면 좋겠어요.

dusdn1702 commented on src/main/java/roomescape/reservation/dao/JdbcReservationDao.java:
```diff
@@ -0,0 +1,62 @@
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+import roomescape.reservation.Reservation;
+import roomescape.time.ReservationTime;
```
reservation과 time으로 패키지를 분리해주셨는데요, time 객체가 reservation 에서 사용이 되면 패키지 분리가 의도대로 동작하는 것 맞을까요?

kjyyjk replied:
`web`, `dao`와 같이 패키지를 나눈 것은 계층 간 의존성을 관리하기 위해서였습니다.
하지만 `reservation`, `time`과 같이 도메인 객체를 중심으로 패키지를 나눈 것은 보기 편해서라는 이유가 강했습니다.

즉 기존에는 계층 간의 의존성만 적절하게 관리되면 상관없을 것이라고 생각했어서 의도대로 동작했다고 보았습니다..!

관련해서 찾아봤을 때 보통 독립적인 도메인들을 패키지 분리해서 도메인 간에도 의존 및 변경 전파를 최소화하는 것 같아요.
현재 Reservation과 ReservationTime은 독립적이지 않고 강하게 결합되어 있다고 생각합니다.
그런 관점에서는 Reservation과 ReservationTime을 분리하지 않고 하나의 패키지에서 다뤄주는 것이 적절할 것도 같아요!

하지만 기존 방식도 보기 편하다는 이유로 괜찮은 방법이라고 생각되어 고민입니다.

kjyyjk replied:
더 생각해봤는데 기존 구조에서 `time` 패키지 내부 구조를 변경하면 `reservation` 패키지 내의 import문도 전부 변경이 일어날 것 같아요.

지금 당장은 큰 단점으로 느껴지지는 않지만, 이후 배포를 했을 때는 문제가 생길 수 있지도 않을까하고 추측해보았습니다!

dusdn1702 replied:
말씀 주신 것처럼 reservation과 time은 reservation 도메인 내부에서 time을 사용하고 있어서 결합도가 높아 패키지 분리의 이유가 궁금했어요. 도메인 별로 dao, web이 패키지가 다 추가된다면 오히려 복잡도가 올라가진 않을지도 같이 검토해보면 좋겠네요.

>  time 패키지 내부 구조를 변경하면 reservation 패키지 내의 import문도 전부 변경되어 배포 시 문제 발생

이 부분에 대해서 어떻게 변경되어 문제가 되는 것인지 다시 설명해주실 수 있을까요?

kjyyjk replied:
그 부분은 저도 잘 모르고 했던 말 같아서 넘어가주시면 감사하겠습니다..ㅎ

결합도가 높기 때문에 패키지 분리가 불필요하다라는 의견에 동의하여 수정했습니다!
121894ef22675b84c5fcd0652d2d54f5b88edf4a

dusdn1702 commented on src/main/java/roomescape/reservation/dao/JdbcReservationDao.java:
```diff
@@ -0,0 +1,62 @@
+import roomescape.time.ReservationTime;
+
+@Repository
+public class JdbcReservationDao implements ReservationDao {
```
Dao를 어떠한 의미에서 이름 붙여주셨는지 궁금해요! 어떤 역할을 하고 있나요?

kjyyjk replied:
Dao는 데이터베이스를 통한 데이터 저장, 조회와 같은 작업들을 수행하는 책임을 가집니다.
Repository와 Dao 중 어떤 이름을 붙일 지 고민해봤지만, 아직까지 둘의 차이가 명확하지 않다고 느껴졌습니다.
그렇기 때문에 미션의 예시 코드에서 사용되며 일반적으로 통용되는 Dao(Data Access Object)를 붙였습니다.
JdbcReservationDao는 ReservationDao의 구현체이며, 내부적으로 jdbc를 직접 사용하기 때문에 그러한 이름을 붙였습니다.

dusdn1702 commented on src/main/java/roomescape/reservation/dao/JdbcReservationDao.java:
```diff
@@ -0,0 +1,62 @@
+    public Reservation save(Reservation reservation) {
+        String sql = "INSERT INTO reservation(name, date, time_id) VALUES(?, ?, ?)";
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        jdbcTemplate.update(connection -> {
```
Statement를 이용하면 index 관리와 데이터 전달에 관리 비용이 추가될 것 같은데, update 메서드를 사용한 이유가 궁금해요.

kjyyjk replied:
해당 코드에서 update 메서드를 사용한 이유는 keyHolder에 자동 생성된 식별자 값을 받아오기 위함이었습니다.

하지만 말씀주신 것처럼 `?`와 `인덱스`를 기반으로 데이터를 전달하는 것은 실수가 생기기에 아주 좋은 방식이라고 여겨집니다.
특히 insert와 같이 db를 update하는 작업에서 index를 잘못 관리했다가는 식겁할 일이 발생할 수도 있을 것 같아요.😅

이후 `NamedParameterJdbcTemplate`에 대해 학습했습니다.
`NamedParameterJdbcTemplate`는 인덱스가 아닌 파라미터명 기반으로 데이터를 전달할 수 있어,
앞서 언급한 문제들을 해결할 수 있습니다.
또한 keyHolder를 이용해 생성된 기본 키를 가져오는 update 메서드도 훨씬 간편하게 작성할 수 있습니다.

다른 해결 방법인 `SimpleJdbcInsert`와 비교해보았을 때도,
삽입 뿐만 아니라 모든 쿼리에서 파라미터명 기반으로 동작할 수 있다는 점이 큰 장점이라고 생각했습니다.

따라서 `NamedParameterJdbcTemplate`을 사용하도록 Dao 구현을 수정했습니다.
f12b10d3081a1fd00215097a30f6606231694cfe

dusdn1702 replied:
너무 좋네요 👍

dusdn1702 commented on src/main/java/roomescape/reservation/dao/ListBasedReservationDao.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.concurrent.atomic.AtomicLong;
+import roomescape.reservation.Reservation;
+
+public class ListBasedReservationDao implements ReservationDao {
```
테스트에서만 사용되는 객체로 보이는데요, 다르게 관리할 수 있는 방법이 없을까요?

kjyyjk replied:
현재는 테스트에서만 사용되지만, 본래 프로덕션 코드에서 사용하기 위해 생성했던 객체이기 때문에 냅뒀습니다!
하지만 그런 객체가 하나 둘 쌓이다 보면 정작 중요한 객체가 잘 보이지 않을 것 같다고 생각됩니다.

만약 다르게 관리한다면 test.fake 패키지로 옮기면서 클래스 명도 `FakeReservationDao`로 수정할 것 같아요.
또한 Test는 멀티 스레드 환경이 아니기 때문에 AtomicLong도 사용하지 않을 것 같습니다.

현재, 그리고 미래에 사용하지 않을 것 같으니 과감하게 옮기는 것이 좋을까요?

dusdn1702 replied:
사용되지 않는 코드가 프로덕션에 남아있으면 해당 객체를 사용해버리는 위험이 있지 않을까요? 사용되지 않는다면 삭제하는게 좋을 것 같아요!

kjyyjk replied:
수정했습니다!
489a9adbf4d64fb4c6ef90885423a55aa9ad6485

dusdn1702 commented on src/main/java/roomescape/reservation/web/ReservationResponse.java:
```diff
@@ -0,0 +1,21 @@
+        LocalDate date,
+        ReservationTimeResponse time
+) {
+    public static ReservationResponse from(Reservation reservation) {
```
도메인 -> dto 변환의 책임을 dto에게 준 이유가 궁금해요.

kjyyjk replied:
저는 Controller에서 도메인 필드 값을 꺼내와 직접 dto를 생성하기 보다는,
객체에게 메세지를 던지는 것이 더 적절하다고 생각했습니다.

따라서 dto나 도메인 객체에게 변환의 책임을 주기로 했습니다.

만약 도메인 객체가 dto 변환 책임을 가져간다면 도메인이 dto에 직접 의존해야 합니다.
하지만 dto는 데이터를 전달하는 객체로써 쉽게 변경될 수 있습니다.
특히 해당 dto는 클라이언트에게 내려주는 응답의 형식을 그대로 따르고 있습니다.

결국 주 관심사인 도메인이 dto로 인해 변경되는 것을 막기 위해, dto에게 dto 변환 책임을 할당했습니다!

dusdn1702 replied:
좋습니다 👍

dusdn1702 replied:
좋습니다 👍

dusdn1702 commented on src/main/java/roomescape/reservation/web/ReservationRequest.java:
```diff
@@ -0,0 +1,11 @@
+import roomescape.reservation.Reservation;
+import roomescape.time.ReservationTime;
+
+public record ReservationRequest(String name, LocalDate date, Long timeId) {
```
요청에서 빈 값을 보내면 
```
Resolved [org.springframework.http.converter.HttpMessageNotReadableException: JSON parse error: Cannot deserialize value of type `java.lang.Long` from String "시간 선택": not a valid `java.lang.Long` value]
``` 
에러가 발생하고 별다른 유저 액션이 없는데요, 유효성 검증에 대한 처리도 추가해주면 좋겠어요.

dusdn1702 replied:
추가로 name을 입력하지 않은 경우 빈 값이 저장되는데 이 부분도 확인이 필요하겠네요.

kjyyjk replied:
유효하지 않은 값이 들어올 경우 예외를 던지게 끔 수정했습니다.

ReservationTime의 시간 Null 검증 추가(https://github.com/woowacourse/spring-roomescape-admin/pull/291/commits/81f9678c7cb6daa54051ac5a5104a2941d8325e3)
ReservationTime의 필드 검증 추가(https://github.com/woowacourse/spring-roomescape-admin/pull/291/commits/068b1d9c5babda1f4d1938ea6bdeee9c43ca7740)
요청 본문 내 데이터의 타입이 잘못된 경우 예외 처리(c127cefa59a21a6bdeee9d8f8c751960f51a12d7)

dusdn1702 commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -0,0 +1,42 @@
+import java.time.LocalDate;
+import roomescape.time.ReservationTime;
+
+public class Reservation {
```
Reservation은 record로 선언하지 않으신 이유가 있으신가요?

kjyyjk replied:
record는 이번 페어 프로그래밍에서 처음 사용해보았고,
record를 사용하면 보일러 플레이트 코드를 제거하고 높은 신뢰도와 함께 불변 데이터를 다룰 수 있다고 학습했습니다.

결론부터 말씀드리자면 `도메인 객체는 record로 선언하지 않는다`는 제 기준을 세워 보았습니다.

그 이유는 record의 구조가 고정되어 있고, 도메인 객체는 제 주 관심사이기 때문입니다.

record는 점 표기법으로 외부에서 프로퍼티에 직접 접근할 수 있습니다.
이러한 특성은 도메인 객체의 내부 구현을 그대로 외부에 드러내는 행위라고 생각했습니다.
이 경우 내부 구현인 프로퍼티 명을 변경했을 때 외부로 변경이 전파될 수 있습니다.

즉 도메인 객체를 record로 선언하는 것은 캡슐화를 위반할 수 있는 여지를 열어두는 것이라는 것을 강하게 느꼈습니다.

물론 지금도 웬만한 객체들이 모든 필드에 getter를 가지게끔 구현하고 있지만, 이는 제가 제어할 수 있습니다.
하지만 record는 그 `구조가 고정`되어 있기 때문에, 제어할 수 없고 강제된다는 것에 불편함을 느꼈습니다.

따라서 저의 주요 관심사인 도메인에서는 편의를 생각하기보다 `직접 제어`하고 싶은 마음이 크기 때문에 record를 
사용하지 않았습니다.

dusdn1702 replied:
`외부로 변경이 전파` 되는 것을 방지하기 위해 도메인 객체는 class 사용하신 것 좋네요 👍

dusdn1702 reviewed (changes requested):
안녕하세요 레오 
이번 미션도 잘 수행해주셨네요. 고생 많으셨습니다.
코멘트 남긴 부분 확인 부탁드릴게요 !

kjyyjk committed:
feat: ReservationTime의 시간 Null 검증 추가

kjyyjk committed:
feat: ReservationTime의 필드 검증 추가

kjyyjk committed:
feat: 요청 본문 내 데이터의 타입이 잘못된 경우 예외 처리

- 현재 단계에서는 우선 예외를 잡아 IllegalArgumentException으로 던지도록 한다.

kjyyjk committed:
refactor: NamedParameterJdbcTemplate을 사용하도록 리팩터링

- 기존에는 JdbcTemplate을 사용하며 쿼리의 ?와 인덱스로 데이터를 전달했다.
하지만 이는 실수하기 좋은 방식으로 유지 보수에 좋지 않다.

- NamedParameterJdbcTemplate은 sql문 내 파라미터명을 기반으로 데이터를 전달하기
때문에 기존에 발생하는 문제를 해결할 수 있다.

dusdn1702 commented on src/main/java/roomescape/reservation/web/ReservationController.java:
```diff
@@ -1,45 +1,60 @@
         return ResponseEntity.notFound().build();
     }
+
+    @ExceptionHandler(HttpMessageNotReadableException.class)
```
ExceptionHandler를 이용해 예외 처리를 하신 이유가 무엇인가요?

kjyyjk replied:
우선 처리하고자 하는 예외가 HttpConverter에서 발생하는 것을 에러 메세지를 통해 확인했습니다.
따라서 메서드 내부에서 처리할 수 없음을 깨달았습니다.

생각한 해결 방법은 다음과 같이 3개였습니다.
1. filter, interceptor를 활용해 converter가 동작하기 이전에 처리한다.
2. 현재 타입이 맞지 않아 예외가 발생하고 있기 때문에 해당 필드의 타입을 Object로 변경한다.
3. exception handler를 활용한다.

방법 1은 학습 단계적으로도, 용도적으로도 많이 어긋난 방식이라고 생각해서 선택하지 않았습니다.
방법 2는 누가 봐도 아닌 것 같았기에 정말 최후의 수단이라는 생각으로 선택하지 않았습니다.
따라서 남은 방법 3을 선택했습니다.
(피드백을 받은 뒤 `@Validated`와 `BindingResult`가 떠올라 시도 해보았으나, 의도대로 동작하지 않았습니다.)

사실 ExceptionHandler가 컨트롤러에서 던지는 예외를 잡을 수 있는 것을 알고는 있었지만,
HttpConverter에서 발생하는 예외도 잡힐지에 대한 확신은 없었습니다.

실제로 잡을 수 있는 것을 알고 나서는 다음과 같이 추론할 수 있었습니다.
Controller 클래스 상에 HttpConverter의 코드가 보이지는 않지만, Controller 내부에서 동작하는 것이 아닐까?하고요.
여기에는 AOP라는 개념이 들어가 있을 것 같지만, 더 깊이 알아보지는 않았습니다.

kjyyjk replied:
추가로 해당 예외는 ReservationController 뿐만 아니라 모든 Controller에서 범용적으로 다뤄주어도 좋을 것 같아요.
그때는 Controller마다 똑같이 `ExceptionHandler`를 이용하거나, `ControllerAdvice`에 대해 학습하고 사용해볼 수 있을 것 같습니다.

dusdn1702 replied:
넵 지금은 적용해본 경험으로 충분할 것 같고, 추후에 더 깊이있게 말씀 주신 내용들을 학습해봐도 좋을 것 같아요.
고생 많으셨습니다 👍

dusdn1702 commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -0,0 +1,63 @@
+import roomescape.time.ReservationTime;
+
+public class Reservation {
+    private final Long id;
```
도메인 객체에서 db로부터 저장된 id값을 가지고 있는 것에 대해 어떻게 생각하시는지 궁금해요!

kjyyjk replied:
처음에는 db 상관 없이 id 필드를 다른 필드들과 동일 선상에서 바라보려고 했습니다.
하지만 미션을 진행할 수록 id 필드가 db에 종속적인 것은 부인할 수 없겠더라고요.

현재처럼 도메인 객체가 db의 키 식별자를 가지고 있으면 여러 문제가 생길 수도 있을 거 같습니다.
예를 들어 키 식별자가 변경 될 경우, 도메인까지 변경이 전파되는 것처럼요..!

이에 대해서는 도메인과 db 사이에 dto를 두어, 도메인 내부에서 db 종속적인 부분을 분리할 수 있을 것 같습니다.
그렇게 한다면 계층 간 격리성을 높이고 변경에 대한 유연성을 확보할 수 있다는 장점이 있을 것 같아요.
하지만 관리해야 할 중간 계층이 늘어나면서 복잡도가 증가한다는 단점도 있을 것 같습니다.

이 부분은 트레이드 오프라고 생각하고, 저라면 기존 방식을 선택할 것 같아요.
현재 요구사항에서는 그렇게까지 복잡도를 가져가며 도메인과 db를 분리했을 때의 이점이 크지 않다고 생각합니다.
하지만 이후 애플리케이션이 비대해지고 요구사항이 복잡해진다면 유지보수를 고려해 분리해볼 수 있을 것 같습니다.

dusdn1702 replied:
넵 저도 공감해요 👍 
다만 id라는 필드를 무의식적으로 당연스레 도메인에 두지는 않았으면 좋겠다고 생각해서 한번 여쭤봤어요.
많이 고민해주셔서 감사해요!

dusdn1702 commented on src/main/java/roomescape/time/dao/JdbcReservationTimeDao.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public ReservationTime save(ReservationTime reservationTime) {
+        KeyHolder keyHolder = new GeneratedKeyHolder();
```
keyHolder를 꼭 사용해야 하는지 궁금해요!

kjyyjk replied:
keyHolder를 사용하지 않고 생성된 id를 가져오려면 `SimpleJdbcInsert#executeAndReturnKey`를 사용해볼 수 있을 것 같아요.

하지만 SimpleJdbcInsert를 사용해야하는 이유는 잘 모르겠습니다..!
삽입 작업을 간편하게 수행할 수 있다지만, 현재 코드랑 비교해본다면 큰 차이가 없는 것 같아요.

또한 sql문을 직접 작성하지 않아도 되기 때문에 안전하다는 점도 있지만,
지금 쿼리가 복잡하지도 않고 피드백 주신 것처럼 Dao에 대한 테스트를 추가한다면 괜찮지 않을까 싶습니다.

혹시 제가 놓치고 있는 부분이나 SimpleJdbcInsert를 사용해야 하는 또 다른 이유가 있을까요?
말씀주시면 더 고민하고 학습 해보겠습니다!

dusdn1702 replied:
말씀 주신 KeyHolder를 추가하지 않아도 된다는 점과 간편한 쿼리 작성 외에 추가적인 내용은 없을 것 같아요.
해당 내용을 인지하고 계신걸로 충분해보입니다. 👍

dusdn1702 commented on src/test/java/roomescape/time/web/ReservationTimeControllerTest.java:
```diff
@@ -0,0 +1,83 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+
+class ReservationTimeControllerTest {
```
Dao 에 대한 단위 테스트는 없을까요??

kjyyjk replied:
제가 작성한 sql문에 이상이 있을 수 있으므로 Dao도 단위 테스트를 수행해주면 좋을 것 같아요.

대신 제가 확인하고 싶은 것은 `Dao가 잘 동작하는가`이고, 이 과정에서 스프링 컨테이너는 불필요하다고 판단했습니다.
따라서 스프링이 제공하는 임베디드 DB를 이용해 Dao 단위 테스트를 작성해보았습니다!
2b87005a94c4ce0017a0d6eed58b7d90ea508821

dusdn1702 replied:
좋습니다. 그렇다면 JdbcDB에 대한 단위 테스트는 없어도 될까요?
embeddedDB와 jdbcDB에 우리가 몰랐던 쿼리의 다른 동작이 있다면 이 테스트로는 알기 어렵지 않을까요?

dusdn1702 reviewed (changes requested):
안녕하세요 레오!
코멘트 남겨주신 것 확인했습니다. 깊이 있게 고민해주셔서 감사해요. 고생 많으셨습니다 :)
몇가지 코멘트 한번 더 확인 부탁드릴게요~

kjyyjk committed:
refactor: ListBasedReservationDao를 테스트 패키지로 이동

- 프로덕션에서는 사용하지 않고 테스트에서만 Fake 객체로 사용되기 때문에 테스트 패키지로 이동했다.
클래스명도 FakeReservationDao로 수정하였다.
테스트에서는 동시성 문제를 고려하지 않아도 되므로 AtomicLong이 아닌 long으로 index를 관리한다.

kjyyjk committed:
refactor: time 관련 클래스들을 reservation 패키지로 이동

- ReservationTime과 Reservation은 서로 강결합되어 있기 때문에
같은 패키지에서 보관하도록 수정한다.

kjyyjk committed:
test: reservation/reservationTime Dao 단위 테스트

- 각 dao가 잘 동작하는 지 테스트한다.
dao를 테스트하는데 스프링 컨텍스트는 필요하지 않다고 생각했다.
따라서 `@SpringBootTest`나 `@JdbcTest`를 사용하지 않았다.
대신 스프링에서 지원하는 임베디드 DB를 사용해 테스트했다.

- 매 테스트 이전에 임베디드 db를 새로 구성하는 것 같지만 testdb라는 이름의 db를 그대로 사용한다.
테스트 환경을 독립적으로 만들기 위해  `generateUniqueName(true)`를 사용한다.
이를 사용하면 매번 다른 이름의 db를 구성한다.

- 각 테스트가 종료한 후 EmbeddedDatabase#shutdown을 호출해 자원을 반납한다.

kjyyjk commented on src/test/java/roomescape/reservation/dao/JdbcReservationTimeDaoTest.java:
```diff
@@ -0,0 +1,99 @@
+import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseType;
+import roomescape.reservation.ReservationTime;
+
+class JdbcReservationTimeDaoTest {
+    private EmbeddedDatabase database;
+    private JdbcReservationTimeDao jdbcReservationTimeDao;
+
+    @BeforeEach
+    public void init() {
+        database = new EmbeddedDatabaseBuilder()
+                .setType(EmbeddedDatabaseType.H2)
+                .generateUniqueName(true)
+                .addScript("schema.sql")
+                .build();
+        jdbcReservationTimeDao = new JdbcReservationTimeDao(new NamedParameterJdbcTemplate(database));
+    }
+
+    @AfterEach
+    public void destroy() {
+        database.shutdown();
+    }
```
Dao를 단위 테스트하는데 스프링 컨테이너는 불필요하다고 판단해, 스프링에서 제공하는 임베디드 DB를 사용했습니다.

테스트 간 독립성을 지키기 위해서는 테스트마다 테이블을 새로 생성하거나, db를 새로 만들어야 했습니다.

저는 각 테스트 환경을 완전히 분리하기 위해 db를 새로 구성해주기를 선택했습니다.
(롤백이나 truncate를 할 수도 있었지만, 시퀀스는 직접 초기화해주어야 한다는 것을 확인해 아예 새로 만드는 방식으로 선택했습니다.)

이때 EmbeddedDatabase는 별도 지정을 해주지 않으면 `testdb`라는 같은 이름의 임베디드 db를 공유하기 때문에,
`.generateUniqueName(true)`로 매 테스트마다 다른 db를 구성하도록 했습니다.

저는 좋은 테스트 방식이라고 생각하는데, 샐리는 어떻게 생각하시는지 궁금합니다!

dusdn1702 replied:
[해당 코멘트](https://github.com/woowacourse/spring-roomescape-admin/pull/291#discussion_r2065649974)에 이어서 말씀드리자면, 동일한 DB이지만 운영 환경과 다른 설정이 존재할 수 있는 부분인데요.
사실은 지금의 구조도 좋아보여요. 지금은 Dao에 대한 테스트를 하고 싶었기 때문에 embedded 를 이용한 것이고, DB 단 쿼리에 대한 테스트를 하고 싶으면 이전에 DM으로 얘기 나눈 것처럼 직접 `@JdbcTest`를 사용하면 된다고 생각해요.
다양한 고민을 통해 실험해보고, 적용하시느라 고생 많으셨습니다 👍 👍

dusdn1702 commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -0,0 +1,79 @@
+    private final ReservationTime time;
+
+    public Reservation(Long id, String customerName, LocalDate reservationDate, ReservationTime time) {
+        validateCustomerName(customerName);
```
유효성 검증을 도메인의 생성자에서 적용하신 이유가 궁금해요!

dusdn1702 commented on src/main/java/roomescape/reservation/ReservationTime.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Objects;
+
+public class ReservationTime {
+    private final Long id;
```
Wrapper type을 이용한 이유는 무엇인가요??

dusdn1702 commented on src/main/java/roomescape/reservation/ReservationTime.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @Override
+    public boolean equals(Object object) {
```
equals & hashCode는 왜 필요했을까요??

dusdn1702 reviewed (approved):
안녕하세요 레오~
피드백 반영해주신 부분들 확인했습니다.
많은 고민을 해주신게 느껴졌어요. 고생 많으셨습니다.
충분히 미션을 진행해주신 것 같아 이번 미션은 여기서 마무리하도록 할게요. 
몇가지 코멘트 추가로 남긴건 확인만 해주시면 될 것 같아요.
다음 미션도 화이팅입니다!
